src/futures.rs

```diff
@@ -0,0 +1,148 @@
+use futures::stream::Stream;
+use std::pin::Pin;
+use std::sync::{mpsc, Arc, Mutex};
+use std::task::{self, Poll, Waker};
+use std::thread;
+
+struct SharedState<T: Send + 'static> {
+    result: Option<Option<T>>,
+    kill: bool,
+    waker: Option<Waker>,
+}
+
+pub struct ThreadedReceiver<T: Send + 'static> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T: Send + 'static> ThreadedReceiver<T> {
+    pub fn new(recv: mpsc::Receiver<T>) -> ThreadedReceiver<T> {
+        let shared_state = Arc::new(Mutex::new(SharedState {
+            result: None,
+            kill: false,
+            waker: None,
+        }));
+
+        // Clone everything to avoid lifetimes
+        let thread_shared_state = shared_state.clone();
+        thread::spawn(move || {
+            loop {
+                let result = recv.recv();
+
+                {
+                    let mut shared_state = thread_shared_state
+                        .lock()
+                        .expect("ThreadedFuture shared state shouldn't be poisoned");
+
+                    if let Ok(result) = result {
+                        shared_state.result = Some(Some(result));
+                    } else {
+                        break;
+                    }
+                }
+
+                // Don't attempt to recv anything else until consumed
+                loop {
+                    let mut shared_state = thread_shared_state
+                        .lock()
+                        .expect("ThreadedFuture shared state shouldn't be poisoned");
+
+                    if shared_state.kill {
```

Comment:
This gets set to true by the `impl Drop` for `ThreadedReceiver`. This allows us to terminate this thread if the consumer doesn't need to consume the entirety of the channel.